code/exercises/camera_calibration/calibrated_lane_detector.py

```diff
@@ -0,0 +1,98 @@
+import numpy as np
+from ..lane_detection.lane_detector import LaneDetector
+from ..lane_detection.camera_geometry import CameraGeometry
+
+
+def get_intersection(line1, line2):
+    m1, c1 = line1
+    m2, c2 = line2
+    #TODO: find intersection of the line.
+    raise NotImplementedError
+
+def get_py_from_vp(u_i, v_i, K):
+    #TODO compute pitch and yaw given the camera intrinsic matrix and vanishing point.
+    raise NotImplementedError
+    return pitch, yaw
+
+class CalibratedLaneDetector(LaneDetector):
+    def __init__(self, calib_cut_v = 200, cam_geom=CameraGeometry(), model_path='./fastai_model.pth'):
+        # call parent class constructor
+        super().__init__(cam_geom, model_path)
+
+        self.calib_cut_v = calib_cut_v
+
+        # build u,v grid
+        uv = []
+        for v in range(calib_cut_v, self.cg.image_height):
+            for u in range(self.cg.image_width):
+                uv.append(np.array([u,v]))
+        self.uv_grid = np.array(uv)
+
+        self.estimated_pitch_deg = 0
+        self.estimated_yaw_deg = 0
+        self.update_cam_geometry()
+        self.mean_residuals_thresh = 2.5 #TODO: adjust this thresh hold to avoid calibration process at curves.
+        self.pitch_yaw_history = []
+        self.calibration_success = False
+
+    def run_and_viz(self, image):
+        _, left_probs, right_probs = self.detect(image)
+        line_left  = self._fit_line_v_of_u(left_probs)
+        line_right = self._fit_line_v_of_u(right_probs)
+        if (line_left is not None) and (line_right is not None):

```

Comment:
I think this "if block" could be written by the student. We could add a comment:
```python
# TODO: If both `line_left` and `line_right` are not None, try to compute the vanishing point using your `get_intersection` function. 
# Then compute pitch and yaw from the vanishing point
# Finally store the pitch and yaw values in `self.pitch_yaw_history`.
# This `run_and_viz` function will be called again and again over time.
# Once enough data is gathered in `self.pitch_yaw_history`, compute mean values for pitch and yaw and store them in ` self.estimated_pitch_deg`and ` self.estimated_yaw_deg` 
# Finally call `update_cam_geometry()` so that the new estimated values are being used
```
The function `add_to_pitch_yaw_history` would then also be deleted in this file